File: mod/common/java/xen/mod/core/Critic.java
package xen.mod.core;

/** TD critic with a dueling head: striatum (reward) or amygdala (harm). Port of xen/brain/critic.py. */
public final class Critic {
	public Mlp net, target;
	public final float gamma, tau;
	final int actions;

	public Critic(int obs, int actions, int[] hidden, float lr, float gamma, float tau, long seed) {
		int[] sizes = new int[hidden.length + 2];
		sizes[0] = obs;
		System.arraycopy(hidden, 0, sizes, 1, hidden.length);
		sizes[sizes.length - 1] = actions + 1;
		this.net = new Mlp(sizes, lr, seed);
		this.target = net.copy();
		this.gamma = gamma;
		this.tau = tau;
		this.actions = actions;
	}

	static float[] q(float[] out) {
		int n = out.length - 1;
		float mean = 0;
		for (int i = 1; i <= n; i++) mean += out[i];
		mean /= n;
		float[] q = new float[n];
		for (int i = 0; i < n; i++) q[i] = out[0] + out[i + 1] - mean;
		return q;
	}

	public float[] values(float[] obs) {
		return q(net.predict(obs));
	}

	public float[][] values(float[][] obs) {
		float[][] out = net.forward(obs, false);
		float[][] qs = new float[obs.length][];
		for (int i = 0; i < obs.length; i++) qs[i] = q(out[i]);
		return qs;
	}

	/** One TD step: Q(s,a) <- signal + gamma^steps * Q_target(s', a'). */
	public void learn(float[][] obs, int[] a, float[] signal, float[][] next, float[] done, int[] nextA, float[] steps) {
		int n = a.length;
		float[][] tOut = target.forward(next, false);
		float[][] out = net.forward(obs, true);
		float[][] grad = new float[n][actions + 1];
		for (int s = 0; s < n; s++) {
			float qNext = q(tOut[s])[nextA[s]];
			double discount = Math.pow(gamma, steps == null ? 1 : steps[s]);
			float y = (float) (signal[s] + discount * (1 - done[s]) * qNext);
			float g = Mlp.huber(q(out[s])[a[s]] - y) / n;
			grad[s][0] = g;
			for (int j = 0; j < actions; j++) grad[s][j + 1] = -g / actions;
			grad[s][a[s] + 1] += g;
		}
		net.backward(grad);
		target.softUpdate(net, tau);
	}
}

File: xen/brain/critic.py
"""Temporal-difference critics.

The same machinery learns two different things:
  * the striatum learns how much *reward* each action leads to;
  * the amygdala learns how much *harm* each action leads to (fear).

Both use a dueling head (a value for the situation plus an advantage per
action), which makes it much easier to learn how actions differ when most of
them lead to similar futures.
"""
import numpy as np

from .nn import MLP, huber_grad


class Critic:
    def __init__(self, obs_dim, n_actions, hidden=(256, 256), lr=3e-4,
                 gamma=0.97, tau=0.01, seed=0):
        self.n_actions = n_actions
        self.net = MLP([obs_dim, *hidden, n_actions + 1], lr=lr, seed=seed)
        self.target = self.net.copy()
        self.gamma = gamma
        self.tau = tau

    @staticmethod
    def _q(out):
        value, advantage = out[:, :1], out[:, 1:]
        return value + advantage - advantage.mean(1, keepdims=True)

    def values(self, obs):
        return self._q(self.net.predict(np.atleast_2d(obs)))

    def learn(self, obs, actions, signal, next_obs, done, next_actions, steps=None):
        """One TD step: Q(s,a) <- signal + gamma^steps * Q_target(s', a')."""
        rows = np.arange(len(actions))
        discount = self.gamma ** (1.0 if steps is None else steps)
        q_next = self._q(self.target.predict(next_obs))[rows, next_actions]
        y = signal + discount * (1.0 - done) * q_next
        out = self.net.forward(obs)
        q = self._q(out)
        diff = q[rows, actions] - y
        g = huber_grad(diff) / len(actions)
        grad = np.zeros_like(out)
        grad[:, 0] = g                                   # dQ/dV = 1
        grad[:, 1:] = -g[:, None] / self.n_actions       # dQ/dA_j = [j == a] - 1/n
        grad[rows, actions + 1] += g
        self.net.backward(grad)
        self.target.soft_update(self.net, self.tau)
        return float(np.abs(diff).mean())
